Account memory of BlockBasedTable::Rep + TableProperties of BlockBasedTableReader

hx235 opened the pull request:
**Summary:**
To add

**Test:**
New unit tests
To add db_bench/stress test/heap profiling on small case

hx235 commented on memory/cache_capacity_based_memory_allocator.h:
```diff
@@ -0,0 +1,71 @@
+
+// CacheCapacityBasedMemoryAllocator is for allocating memory based on available
+// cache space by charging the memory to the cache.
+// For now, it does not support the actual memory allocation and only the
+// memory charging.
```
[Open for discussion]
For BlockBasedTableReader, memory usages are allocated somewhere outside of this Allocator and tracking-only is sufficient. Therefore I didn't impl the conventional Allocate/Deallocate function. But for future use-case of `CacheCapacityBasedMemoryAllocator`, we might (?) need actual allocation.

ajkr replied:
OK, so it turns out we are reserving based on `BlockBasedTable::ApproximateMemoryUsage()`, not making (de)allocations go through the `Allocator` interface. Do you think we should just use the CacheReservationManager interface directly (with appropriate locking)?

hx235 replied:
> not making (de)allocations go through the Allocator interface

hx235 replied:
> Do you think we should just use the CacheReservationManager interface directly (with appropriate locking)?

Yeah .. I've been thinking about this too.  This task deepened my understanding toward Allocator approach - the difficulties of it lies in going through explicit (de)allocation through our `Allocator` interface for objects whose de(allocation) are handles automatically (like [new_table_properties->user_collected_properties.insert({key, raw_val.ToString()});](https://github.com/facebook/rocksdb/blob/7.0.fb/table/meta_blocks.cc#L336-L337)).
If we were not able to achieve that, then "Allocator" is essentially a tracker.


> CacheReservationManager interface directly (with appropriate locking)?

I am thinking of making `CacheReservationManager` optionally internal thread-safe (`new CacheReservationManager(cache,  true \*enable_internal_locking*\)`). "Internal thread-safe" because it's easier to do so than passing the lock pointer around if we were to do so for this PR without Allocator layer. "Optionally" is because the `WriteBufferManager` maintains a atomic-relaxed `memory_used_` that serves both as a stat for query without locking (trade accuracy for speed) and a referenced data point for charging and releasing cache entry (requires accuracy) and I have no plan to deal with the refactoring issue there (double locking and deadlock) for now.

hx235 replied:
> It seems achievable but could take a long time. For example we could make STL use our allocator like [7b1890f0b3e096a43006e4163b98f6316ac1e91b](https://github.com/ClickHouse/ClickHouse/blob/5e641e2dace33180ec34eb913deaaebd2ad01c18/src/Common/AllocatorWithMemoryTracking.h), 

It's true - I looked into that route too but I agreed with what you said about "take a long time" but worth exploring when we have time and are bored :).


> Can you make a thread-safe wrapper or something?
Aha - wrapper also works (just haven't thought of that)!

hx235 replied:
Done

hx235 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -79,6 +80,11 @@ extern const std::string kHashIndexPrefixesBlock;
 
 BlockBasedTable::~BlockBasedTable() {
+  std::size_t mem_usage = ApproximateMemoryUsage();
+  if (rep_->table_reader_mem_allocator) {
```
TODO: replace with CacheReservationHandle approach

hx235 commented on table/block_based/block_based_table_factory.cc:
```diff
@@ -419,6 +423,12 @@ BlockBasedTableFactory::BlockBasedTableFactory(
   RegisterOptions(&table_options_, &block_based_table_type_info);
+
+  if (table_options_.reserve_table_reader_memory &&
+      table_options_.no_block_cache == false) {
```
[Note] A fun little thing I found is when `table_options_.no_block_cache` is set to 1 by option string, the negate of it `!table_options_.no_block_cache` in dbg build (but not DEBUG_LEVEL=1) evaluates to `True` incorrectly ......

hx235 commented on table/meta_blocks.cc:
```diff
@@ -347,32 +349,46 @@ Status ReadTablePropertiesHelper(
+      new_table_properties->IncreaseApproximateMemoryUsage(sizeof(val));
     } else if (key == TablePropertiesNames::kDbId) {
       new_table_properties->db_id = raw_val.ToString();
+      new_table_properties->IncreaseApproximateMemoryUsage(raw_val.size());
```
Not my favorite code repetition either but couldn't find an easy way to avoid that ...

hx235 replied:
Will fix this (previously I avoided it due to the nested if-else but now I think it is still better than code duplication)

hx235 replied:
I think one "longer-term" fix is to avoid "key ==  TablePropertiesNames::XX" pattern and use set. Opened a tech debt for it.

hx235 replied:
Done by on-the-fly memory approximation for table properties

ajkr reviewed (changes requested):
Thanks very much for doing this work. Left a few suggestions/questions.

hx235 commented on table/block_based/block_based_table_reader_test.cc:
```diff
@@ -249,6 +250,259 @@ TEST_P(BlockBasedTableReaderTest, MultiGet) {
+    options_.table_factory.reset(NewBlockBasedTableFactory(table_options));
+  }
+
+  void CreateTable(const std::string& table_name) {
```
Got you - let me polish the test a bit to make it less duplicated and more readable.

hx235 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -79,6 +80,11 @@ extern const std::string kHashIndexPrefixesBlock;
 extern const std::string kHashIndexPrefixesMetadataBlock;
 
 BlockBasedTable::~BlockBasedTable() {
+  std::size_t mem_usage = ApproximateMemoryUsage();
```
Good point - using some kind of CacheReservationHandle eliminates the need to inquiry ApproximateMemoryUsage() again during destructor. Will consider that.

hx235 replied:
Improved along using CRM. Done.

hx235 commented on include/rocksdb/table.h:
```diff
@@ -300,12 +300,32 @@ struct BlockBasedTableOptions {
+  // Charged memory usage includes:
+  // 1. Table properties
+  // 2. Some internal data structures
+  // 3. More to come...
```
I originally left those out cuz the implementation of those ApproximateMemoryUsage() just included basic sizeof(class), which seemed to be an underestimation to me. But now when I think about it again, those seems to be a fine estimate of i

hx235 replied:
I recalled my previous rationale was the current `BlockBasedTable::ApproximateMemoryUsage()` included `IndexReader`, `FilterBlockReader`, `UncompressionDictReader`, which are non-public headers (unlike TableProperties). So I decided to include them under "internal data structures".  

But reading your comments makes me realize maybe it's clearer to specifically state them, especially we have a feature "cache_index_and_filter_blocks == false". Will improve this API doc.

hx235 replied:
Done

hx235 commented:
Waiting for CI signals 
Update:
- Clarified BlockBasedTableOptins::reserve_table_reader_memory API doc
- Refactored block_based_table_reader_test, including refactoring out BlockBasedTableReaderTestBase, encapsulating details on approximating table reader mem, removing a unused function and minor code changes 
- Used CacheReservationManagerThreadsafeWrapper::CacheReservationHandle for BlockBasedTableReader's reservation for better RAII on cache reservation
  - - This includes modifying CacheReservationManager and its API comment to add better support for multi-threaded usage of CacheReservationHandle; renamed usage of previous CacheReservationManagerHandle
  - - Removed unused class CacheCapacityBasedMemoryAllocator

hx235 commented on table/block_based/block_based_table_reader_test.cc:
```diff
@@ -249,6 +250,259 @@ TEST_P(BlockBasedTableReaderTest, MultiGet) {
+
+  std::size_t updated_max_table_reader_num =
+      BlockBasedTableReaderCapMemoryTest::
+          ApproximateMaxTableReaderNumBeforeCacheFull(cache_capacity / 2,
```
See above.

hx235 replied:
To understand this argument better, are you suggesting that disabling table reader reservation in approximating the table reader memory does not give an convincing estimate of the memory due to non-obvious side effect?

(One thing for simply returning Status::OK() while table reader reservation is *enable* is that releasing those reservation as part of the features does not have the right cache handles any more)

ajkr replied:
> For me to understand this argument better, are you suggesting that disabling table reader reservation in approximating the table reader memory does not give a convincing estimate of the memory due to non-obvious side effect?

The code is different now but I thought that table reader memory reservation was disabled after the ApproximateMaxTableReaderNumBeforeCacheFull() here, and there wasn't any code to reenable it - even in the case where the test parameter was true.

hx235 replied:
> was disabled by the ApproximateMaxTableReaderNumBeforeCacheFull() here, and there wasn't any code to reenable it - even in the case where the test parameter was true.

Got you - yeah the previous version of my test was not well written for readability. Let's see how the improved version read like :)

> (One thing for simply returning Status::OK() while table reader reservation is enable in memory approximation is that releasing those reservation as part of the feature now does not have the right cache handles)

I was wrong about this - I can change the release to fix this too. But it's irrelevant now as mentioned above.

ajkr reviewed (commented):
Do you mind importing it? Sorry it's hard to read on GitHub.

Replies to outdated comments:

> Not sure how "disable the cache Insert() while that particular table is created." is done but updated the test to use ConfigureTableFactory() and added more comment. Let's see if this helps.

I meant since you are the implementor of the cache, you could make Insert() return Status::OK() without doing any work while you calculate per-table memory.

hx235 commented:
Sure - although I need to fix the ci/circleci: build-linux-clang10-clang-analyze. Importing it

hx235 commented:
> Do you mind importing it? Sorry it's hard to read on GitHub.
> 
> Replies to outdated comments:
> 
> > Not sure how "disable the cache Insert() while that particular table is created." is done but updated the test to use ConfigureTableFactory() and added more comment. Let's see if this helps.
> 
> I meant since you are the implementor of the cache, you could make Insert() return Status::OK() without doing any work while you calculate per-table memory.

Got you - let me try that too (and looking into the ci/circleci: build-linux-clang10-clang-analyze )

ajkr commented on cache/cache_reservation_manager.h:
```diff
@@ -132,7 +154,41 @@ class CacheReservationManager
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle<R>>
+          *handle);
+
+  // One of the two ways of releasing cache reservation, used with
+  // MakeCacheReservation(). It's usually called on the destruction of a
+  // CacheReservationManager::CacheReservationHandle<R> handle, which achieves
+  // better RAII per cache reservation.
+  // See CacheReservationManager::UpdateCacheReservation()
+  // for the other way.
+  //
+  // Use ONLY one of them to prevent unexpected behavior.
+  //
+  // Release dummy entries in the cache for the incremental memory usage
+  // to match the size of total dummy entries with the least multiple of
+  // kSizeDummyEntry greater than or equal to the total memory used.
+  //
+  // WARNING: The passed-in incremental_memory_used should be resulted from
+  //          calling GetIncrementalMemoryUsed() on some
+  //          CacheReservationManager::CacheReservationHandle<R>> object
+  //          created by MakeCacheReservation() and hasn't had its cache
+  //          reservation released
+  //
+  // @param incremental_memory_used The number of bytes previously increased in
+  //        memory usage.
+  //
+  //        Calling GetTotalMemoryUsed() afterward will return the total memory
+  //        decreased by this number.
+  //
+  //        Since the class is NOT thread-safe, external synchronization in
+  //        calling ReleaseCacheReservation() is needed if you want
+  //        GetTotalMemoryUsed() indeed returns the latest memory used.
+  //
+  // @return It always returns Status::OK()
+  template <CacheEntryRole R>
+  Status ReleaseCacheReservation(std::size_t incremental_memory_used);
```
Have you tried making this private (or protected if you follow another suggestion -- see below)? the inner class might be able to access a private function of the outer class even without declaring it a friend.

hx235 replied:
> the inner class might be able to access a private function of the outer class even without declaring it a friend.

Yeah - the issue I ran into is, unlike Java, C++ needs to store a raw pointer or reference of the enclosing class in child class https://stackoverflow.com/questions/486099/can-inner-classes-access-private-variables.

I tried to store smart pointer (std::shared_ptr) instead and couldn't get it work ...

ajkr replied:
The error I got when making it private is in the access from CacheReservationManagerThreadSafeWrapper::ReleaseCacheReservation, which isn't an inner class. After working around that it seemed to work.

hx235 replied:
Done by making ReleaseCacheReservation protected as suggested.

ajkr commented on cache/cache_reservation_manager.h:
```diff
@@ -92,8 +111,11 @@ class CacheReservationManager
-  // One of the two ways of reserving/releasing cache,
-  // see CacheReservationManager::UpdateCacheReservation() for the other.
+  // One of the two ways of reserving cache, used with
+  // ReleaseCacheReservation().
```
Does the client need to know about ReleaseCacheReservation()? I thought the client deletes the handle.

hx235 replied:
See above.

hx235 replied:
Done - see above.

ajkr commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +228,75 @@ class CacheReservationManager
+  // This class is NOT thread-safe
+  //
+  // See CacheReservationManager::CacheReservationManager for more API usage.
+  template <CacheEntryRole R>
+  class CacheReservationHandle {
+   public:
+    // REQUIRES: cache_res_mgr_threadsafe_wrapper != nullptr
+    explicit CacheReservationHandle(
+        std::size_t incremental_memory_used,
+        std::shared_ptr<CacheReservationManagerThreadSafeWrapper>
+            cache_res_mgr_threadsafe_wrapper);
+
+    ~CacheReservationHandle();
+
+    std::size_t GetIncrementalMemoryUsed() { return incremental_memory_used_; }
+
+   private:
+    std::size_t incremental_memory_used_;
+    std::shared_ptr<CacheReservationManagerThreadSafeWrapper>
+        cache_res_mgr_threadsafe_wrapper_;
+  };
```
Wait.. Now there are two `CacheReservationHandle` classes. How about making CacheReservationManagerThreadSafeWrapper a subclass of CacheReservationManager (alternative name idea: "ConcurrentCacheReservationManager") instead of a wrapper so we only need one?

hx235 replied:
Interesting .. I was trying a similar thing to consolidate two CacheReseravationHandle but couldn't quite remember what didn't work. Let me try again harder

ajkr replied:
As a subclass I'd just implement the methods like (i.e., not as a wrapper anymore):

```
... ConcurrentCacheReservationManager::ReleaseCacheReservation(...) {
  std::lock_guard<std::mutex> lock(mu_);
  return CacheReservationManager::ReleaseCacheReservation(...);
}
```

hx235 replied:
IIUC, ReleaseCacheReservation is now protected so in a ConcurrentCacheReservationManager object, when trying to do something like cache_res_mgr_->ReleaseCacheReservation on the stored object cache_res_mgr_, it can not access the protected member of that object cache_res_mgr_. 

I have to duplicate the impl of CacheReservationManagerImpl::ReleaseCacheReservation ConcurrentCacheReservationManager to get around with this unfortunately. See note on the PR (attaching it ..)

ajkr commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +228,75 @@ class CacheReservationManager
+  // Same as CacheReservationManager::CacheReservationHandle
+  // expect that it's for CacheReservationManagerThreadSafeWrapper
+  //
+  // This class is NOT thread-safe
```
It seems to be thread safe as you can count on constructor and destructor being called from a single thread before and after all other threads use it, respectively.

hx235 replied:
So we don't need to concern about the case where thread A is deleting the handle while thread B is trying to read it? Although I don't really know how to solve this if it's something we should concern about to ever make the class thread-safe. 

Anyway just a thought and I can remove the thread-safe warning as suggested.

ajkr replied:
> So we don't need to concern about the case where thread A is deleting the handle while thread B is trying to read it? 

Right, other threads shouldn't call member functions while the destructor is running. I don't think there's any way for that to be safe.

hx235 replied:
Done

hx235 commented:
Update: waiting for CI signals to be ready for review
- Addressed feedback above
- Move CRM's template from function level to class level 
  - In additionally, I de-templated CRM to get around with the WBM header issue: forward declaration of a template class (i.e, templated CRM) with an incomplete type (as CacheEntryRole as we don't want to include it in public header WBM)
  - We now have CRM as an interface and CRMImpl (templated implement) and ConcurrentCRM (decorator)
  - Refactored existing usage of CRM
  - Clarified CRM api doc along the way
  - Left one comment of design weirdness.

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +213,61 @@ class CacheReservationManager
+  }
+  inline Status MakeCacheReservation(
       std::size_t incremental_memory_used,
-      std::shared_ptr<CacheReservationManager> cache_res_mgr);
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle> *handle)
+      override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    Status s = cache_res_mgr_->UpdateCacheReservation(
+        cache_res_mgr_->GetTotalMemoryUsed() + incremental_memory_used);
+    (*handle).reset(new CacheReservationHandle(incremental_memory_used,
+                                               shared_from_this()));
 
-  ~CacheReservationHandle();
+    return s;
```
[Note]
 The reason why I didn't do 
```
std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
return cache_res_mgr_->MakeCacheReservation(..);
```

is that the handle needs to point to a ConcurrentCacheReservationManager for synchronized ReleaseCacheReservation on handle's destruction. It feels like a "leak" of opportunity of data race into CRMHandle class out of CRM for the benefit of RAII and now we are paying for that by this code duplication treatment.

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -26,7 +26,8 @@
 // memory used
-class CacheReservationManager {
+class CacheReservationManager :
+      public std::enable_shared_from_this<CacheReservationManager>{
```
[Note] we need to store pointer to CacheReservationManager in CacheReservationHandle so we can't really do "std::unique_ptr<CacheReservationManager>" in call-sites, enforced by compilation error from enable_shared_from_this if do so (hence extra changes included in this PR) 

Benchmark will be included for those changes for concerns on the slow-down from shared_ptr. 

A relevant discussion https://github.com/facebook/rocksdb/pull/9073/files#r748641039 - I can't really think of a way to improve this so I'll leave it as it is right now but welcome any ideas/thoughts!

hx235 commented on table/block_based/block_based_table_reader_test.cc:
```diff
@@ -456,12 +456,14 @@ TEST_P(BlockBasedTableReaderCapMemoryTest, CapMemoryUsageUnderCacheCapacity) {
+    // ApproximateTableReaderMem() and actual table reader mem
     EXPECT_GE(opened_table_reader_num, max_table_reader_num * 0.99);
-    EXPECT_LE(opened_table_reader_num, max_table_reader_num);
+    EXPECT_LE(opened_table_reader_num, max_table_reader_num * 1.01);
```
[Note] this is a response to failure of "EXPECT_LE(opened_table_reader_num, max_table_reader_num);" only https://app.circleci.com/pipelines/github/facebook/rocksdb/12581/workflows/770d0743-6be0-4c8c-af5e-0a831bea3a56/jobs/334853 by `opened_table_reader_num = 820` vs `max_table_reader_num = 816`.

Still debugging it but my guess is `..underestimate max_table_reader_num due to the gap between ApproximateTableReaderMem() and actual table reader mem`.

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -13,54 +13,91 @@
+// CacheReservationManager is an interface for reserving cache space for the
+// memory used
+class CacheReservationManager
+    : public std::enable_shared_from_this<CacheReservationManager> {
```
[Note] we need to store pointer to CacheReservationManager in CacheReservationHandle so we can't really do "std::unique_ptr" in call-sites, enforced by compilation error from enable_shared_from_this if do so (hence extra changes included in this PR)

Benchmark will be included for those changes for concerns on the slow-down from shared_ptr.

A relevant discussion https://github.com/facebook/rocksdb/pull/9073/files#r748641039 - I can't really think of a way to improve this so I'll leave it as it is right now but welcome any ideas/thoughts!

hx235 replied:
> Do we need enable_shared_from_this?

This reason to have "enable_shared_from_this" is more for creating a shared_ptr from *this* pointer (https://github.com/facebook/rocksdb/pull/9748/files#diff-b95042a5d09f06814f920b8c334a4f0aff0574c41981d1fc82a93d1c87e774b5R240). Based on a quick search, extending from enable_shared_from_this and calling shared_from_this is the only standard way of doing so.

ajkr replied:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

ajkr replied:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

ajkr replied:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

hx235 replied:
> It's not necessary to change in this PR anyways.

Okay - then I won't make any immediate decision about this in this PR. 

> But it seems to sacrifice some ability to notice reservation leaks, like in cases where we have test coverage that's not specific to CRM.

But for my future reference (tech-debt kind of thing), can you explain more about how in this case, "enable_shared_from_this" seems to sacrifice some ability to notice reservation leaks?

An alternative is to use raw pointer but maybe there is some other design pattern to explore in future for this.

ajkr replied:
> But for my future reference (tech-debt kind of thing), can you explain more about how in this case, "enable_shared_from_this" seems to sacrifice some ability to notice reservation leaks?

Allowing the handle to extend the manager's lifetime means we could leak the handle until the end of the program assuming there's no CRM-specific test coverage for that case (presumably CRM-specific test coverage would assert memory is reduced when it should be). If there's non-CRM test coverage for that code path we would notice such a leak via ASAN because it'd attempt to access a deleted CRM. Of course the risk is there's no test coverage for the leak path in which case extending the manager's lifetime until the handle is finally destroyed is beneficial.

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +208,59 @@ class CacheReservationManager
       std::size_t incremental_memory_used,
-      std::shared_ptr<CacheReservationManager> cache_res_mgr);
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle> *handle)
+      override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    Status s = cache_res_mgr_->UpdateCacheReservation(
+        cache_res_mgr_->GetTotalMemoryUsed() + incremental_memory_used);
+    (*handle).reset(new CacheReservationHandle(incremental_memory_used,
+                                               this->shared_from_this()));
 
-  ~CacheReservationHandle();
+    return s;
```
[Note]
 The reason why I didn't do 
```
std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
return cache_res_mgr_->MakeCacheReservation(..);
```

is that the handle needs to point to a ConcurrentCacheReservationManager for synchronized ReleaseCacheReservation on handle's destruction. It feels like a "leak" of opportunity of data race into CRMHandle class out of CRM for the benefit of RAII and now we are paying for that by this code duplication treatment.

hx235 replied:
Got you - will try out your suggestion.

hx235 replied:
Okay - now I can see how bad Github UI is. "Got you - will try out your suggestion." was meant to be reply to other threads. And I need to look into your "see below" first.

hx235 replied:
Okay - will still try your suggestion after reading!

hx235 replied:
Good point on making ConcurrentCacheReservationManager::CacheReseravationHandle a decorator of CacheReservationManager::CacheReseravationHandle, which climates two design weirdness I previously pondered. Done.

hx235 commented:
Update:
For benchmarking write buffer manager performance, `./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR -disable_auto_compactions=1 -cost_write_buffer_to_cache=true -write_buffer_size=1048576  | egrep 'fillseq' revealed 5% of regression caused by this PR - needs further investigation.

#-run | (pre-PR) avg | std ms | (post-PR) avg ms | std ms | change (%)
-- | -- | -- | -- | -- | --
10 | 1.7309 | 0.0645019 | 1.8345 | 0.0500045 | 5.985325553
20 | 1.6944 | 0.0415589 | 1.824 | 0.0443712 | 7.648725212
40 | 1.70415 | 0.0577726 | 1.79563 | 0.0746095 | 5.368072059
80 | 1.66829 | 0.046114 | 1.78198 | 0.0539437 | 6.814762421
160 | 1.68704 | 0.0582294 | 1.79145 | 0.055626 | 6.18894632
320 | 1.68486 | 0.0458318 | 1.75307 | 0.0469916 | 4.048407583

hx235 commented:
> ~@ajkr feel free to postpone reviewing as it appears that this PR introduces regression 5% on write. `./db_bench -benchmarks=fillseq -db=/dev/shm/testdb -disable_auto_compactions=1` post-PR is 3% slower than pre-PR. I need to look into this next week.~
> 
> ~For benchmarking write buffer manager performance, `./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR -disable_auto_compactions=1 -cost_write_buffer_to_cache=true -write_buffer_size=1048576 | egrep 'fillseq' revealed 5% of regression caused by this PR - needs further investigation.~
> 
> #-run	(pre-PR) avg micros/op	std micros/op	(post-PR) avg micros/op	std micros/op	change (%)
> 10	1.7309	0.0645019	1.8345	0.0500045	5.985325553
> 20	1.6944	0.0415589	1.824	0.0443712	7.648725212
> 40	1.70415	0.0577726	1.79563	0.0746095	5.368072059
> 80	1.66829	0.046114	1.78198	0.0539437	6.814762421
> 160	1.68704	0.0582294	1.79145	0.055626	6.18894632
> 320	1.68486	0.0458318	1.75307	0.0469916	4.048407583

Update:  @ajkr  Ready for review
Actually, I re-run a benchmark with no flush to benchmark on CRM's interaction in WBM `./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR -disable_auto_compactions=1 -cost_write_buffer_to_cache=true -write_buffer_size=10000000000 | egrep 'fillseq'`. Results seems fine:

#-run | (pre-PR) avg micros/op | std micros/op | (post-PR)  avg micros/op | std micros/op | change (%)
-- | -- | -- | -- | -- | --
10 | 19.6371 | 0.852601 | 19.6494 | 0.430577 | 0.06263654002
20 | 19.6405 | 0.871742 | 19.6847 | 0.332848 | 0.2250451872
40 | 20.0491 | 0.741681 | 20.0872 | 1.00278 | 0.1900334678

ajkr commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +213,61 @@ class CacheReservationManager
+  std::mutex cache_res_mgr_mu_;
   std::shared_ptr<CacheReservationManager> cache_res_mgr_;
+
+  inline Status ReleaseCacheReservation(
+      std::size_t incremental_memory_used) override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    assert(cache_res_mgr_->GetTotalMemoryUsed() >= incremental_memory_used);
+    std::size_t updated_total_mem_used =
+        cache_res_mgr_->GetTotalMemoryUsed() - incremental_memory_used;
+    Status s = cache_res_mgr_->UpdateCacheReservation(updated_total_mem_used);
+    return s;
+  }
```
Hm when complaining "there are two CacheReservationHandle classes" what I really meant to complain is there are two unrelated handle classes. But the handle destructor is the public interface to releasing memory, so it makes sense to have multiple. A ConcurrentCacheReservationManager::CacheReservationHandle could wrap a CacheReservationManagerImpl::CacheReservationHandle and provide locking.

ajkr replied:
btw I tried out this suggestion here in case it's useful - https://github.com/ajkr/rocksdb/commit/87f73be5acb0158ea0c8483dc1497528550d39ca

ajkr replied:
> btw I tried out this suggestion here in case it's useful - ajkr@87f73be

Make sure this is visible since the "Add single comment" appears to not make comments visible in the same way as a full review (?).

ajkr replied:
> btw I tried out this suggestion here in case it's useful - ajkr/rocksdb@87f73be

Make sure this is visible since the "Add single comment" appears to not make comments visible in the same way as a full review (?).

hx235 replied:
Done - similar to your suggestion excepts that (a) no template for  CacheReservationManagerImpl<R>::CacheReservationHandle (b) remove useless access method for std::size_t incremental_memory_used in (c) make CacheReservationManager::CacheReservationHandle more as an interface (i.e, no data stored) to be extended (just like CacheReservationManager)

ajkr replied:
> (a) no template for CacheReservationManagerImpl::CacheReservationHandle

Good finding, I did not know C++ does this.

ajkr reviewed (commented):
> In additionally, I de-templated CRM to get around with the WBM header issue: forward declaration of a template class (i.e, templated CRM) with an incomplete type (as CacheEntryRole as we don't want to include it in public header WBM)

Good idea to only template the CRMImpl subclass to workaround this! I had assumed we'd need to separate WBMImpl from WBM to avoid the forward declaration, and had given up as that seemed like too much refactoring work.

hx235 commented on table/meta_blocks.cc:
```diff
@@ -347,32 +349,46 @@ Status ReadTablePropertiesHelper(
         continue;
       }
       *(pos->second) = val;
+      new_table_properties->IncreaseApproximateMemoryUsage(sizeof(val));
```
Will fix this.

hx235 replied:
Done

hx235 commented on include/rocksdb/table_properties.h:
```diff
@@ -301,6 +301,13 @@ struct TableProperties {
+
+  std::size_t ApproximateMemoryUsage() const;
+
+  void IncreaseApproximateMemoryUsage(std::size_t delta);
```
new_table_properties-> IncreaseApproximateMemoryUsage will be called in ReadTablePropertiesHelper() in meta_blocks.cc.  Are you suggesting to make IncreaseApproximateMemoryUsage private and also friend ReadTablePropertiesHelper() with TableProperties class?

pdillinger replied:
OK, I guess hiding the implementation details is a little trickier. Also, it just occurred to me that we are using more space in TableProperties now to hold the `approx_mem_usage_` even though I suspect we only need to get ApproximateMemoryUsage once. Can't we compute the usage on the fly, perhaps using a pattern like in `TEST_SetRandomTableProperties` to walk over the string properties without lots of copy-paste?

hx235 replied:
Let me look into that.

hx235 replied:
Done with some refactoring and warning notes to such hack. Added to tech debt.

ajkr commented on table/table_properties.cc:
```diff
@@ -213,6 +214,30 @@ TableProperties::GetAggregatablePropertiesAsMap() const {
+#endif  // ROCKSDB_MALLOC_USABLE_SIZE
+
+  std::pair<const std::string*, const std::string*> string_prop_start_end_pos =
+      TEST_GetStringPropStartEndPosition(this);
```
The function is called TEST_* but it doesn't appear to be called only in test code.

hx235 replied:
Done by renaming as  Get*PropStartEndPosition

ajkr commented on table/table_properties.cc:
```diff
@@ -282,24 +307,44 @@ const std::string TablePropertiesNames::kFastCompressionEstimatedDataSize =
 }
 #endif
 
+std::pair<const uint64_t*, const uint64_t*> TEST_GetUint64TPropStartEndPosition(
+    const TableProperties* const_props) {
+  TableProperties* props = const_cast<TableProperties*>(const_props);
+  uint64_t* pu = &props->orig_file_number;
+  assert(static_cast<void*>(pu) == static_cast<void*>(props));
+  const uint64_t* pu_end = reinterpret_cast<const uint64_t*>(&props->db_id);
+  return std::pair<const uint64_t*, const uint64_t*>(pu, pu_end);
+}
+
+std::pair<const std::string*, const std::string*>
+TEST_GetStringPropStartEndPosition(const TableProperties* const_props) {
+  TableProperties* props = const_cast<TableProperties*>(const_props);
+  std::string* ps = &props->db_id;
+  assert(static_cast<void*>(const_cast<uint64_t*>(
+             TEST_GetUint64TPropStartEndPosition(props).second)) ==
+         static_cast<void*>(ps));
+  const std::string* ps_end =
+      reinterpret_cast<const std::string*>(&props->user_collected_properties);
+  return std::pair<const std::string*, const std::string*>(ps, ps_end);
+}
```
This is not easy on the eyes. Do we need it? I thought we were adding the sizes of the strings or something similarly understandable before.

ajkr replied:
OK it makes enough sense to me. Just make sure to eliminate all `TEST_` in non-test code, `const_cast`s, and `reinterpret_cast`s.

hx235 replied:
This is suggested by https://github.com/facebook/rocksdb/pull/9748#discussion_r841971925. Added a [test](https://github.com/facebook/rocksdb/pull/9748/files#diff-d17c3d6f851ee716cc49892ea8712195b0aa032285c654370d29b202d07e4b0bR414) /[comment](https://github.com/facebook/rocksdb/pull/9748/files#diff-5653feed9ef035ce3869aa3c8c2cae720229948bf901638888c05656f7219f08R15) for this too.

Although I still agree it's not easy to understand nor encouraged. I was hoping to name it TEST_XXX to discourage usage of it but I will remove the "TEST_" as suggested.

ajkr replied:
> Although I still agree it's not easy to understand nor encouraged. I was hoping to name it TEST_XXX to discourage usage of it

I think if the C++ is not safe we should actually only use it in tests, not here either. But let's see if there's a proof that `reinterpret_cast<const std::string*>(&props->user_collected_properties)` is guaranteed to be the right string pointer for us to compare against.

hx235 replied:
> What if `user_collected_properties` has different alignment from string?

I believe we aren't reading any value pointed by "reinterpret_cast<const std::string*>(&props->user_collected_properties)" cuz the value won't be the correct as it's reinterpreted as string "incorrectly". We only use the address for purpose of advancing the pointer and the "reinterpret_cast" is only for `ps < ps_end` in the for statement. 

cc @pdillinger as it seems that you started this [hack](https://github.com/facebook/rocksdb/blame/36bc3da97fcb6aa2df6e43df27552f49b762d61e/table/table_properties.cc#L283) and correct me if I'm wrong.

hx235 replied:
> OK it makes enough sense to me. Just make sure to eliminate all TEST_ in non-test code, const_casts, and reinterpret_casts (ideally).

Will do.

hx235 replied:
Done by (a) checking non TEST_ call-sites of GetXXXPropStartEndPosition do not have const_casts, and reinterpret_casts (b) putting all the const_casts in GetXXXPropStartEndPosition in assertion code (c) can't do anything with `reinterpret_cast ` within the GetXXXPropStartEndPosition though

hx235 replied:
> I think if the C++ is not safe we should actually only use it in tests, not here either. But let's see if there's a proof that `reinterpret_cast<const std::string*>(&props->user_collected_properties)` is guaranteed to be the right string pointer for us to compare against.

Is something like this https://github.com/facebook/rocksdb/pull/9748/files#diff-d17c3d6f851ee716cc49892ea8712195b0aa032285c654370d29b202d07e4b0bR434-R435 efficient? Casting both to void* and check address equality

ajkr replied:
> I believe we aren't reading any value pointed by "reinterpret_cast<const std::string*>(&props->user_collected_properties)" cuz the value won't be the correct as it's reinterpreted as string "incorrectly". We only use the address for purpose of advancing the pointer and the "reinterpret_cast" is only for ps < ps_end in the for statement.

Would we dereference an invalid pointer if there were any padding bytes between compression_options and user_collected_properties?

ajkr replied:
For example add:

```
$ git diff include/rocksdb/table_properties.h 
diff --git a/include/rocksdb/table_properties.h b/include/rocksdb/table_properties.h
index 0c91fa854..fa157f7ee 100644
--- a/include/rocksdb/table_properties.h
+++ b/include/rocksdb/table_properties.h
@@ -285,7 +285,7 @@ struct TableProperties {
   std::string compression_options;
 
   // user collected properties
-  UserCollectedProperties user_collected_properties;
+  UserCollectedProperties user_collected_properties alignas(16);
   UserCollectedProperties readable_properties;
 
   // convert this object to a human readable form
```

Under ASAN the test then fails, I assume because ASAN uses nonzero padding bytes and we then interpreted it as a string of some nonzero size.

```
$ ./db_sst_test --gtest_filter=DBSSTTest.OpenDBWithInfiniteMaxOpenFilesSubjectToMemoryLimit
...
db_sst_test: cache/cache_reservation_manager.cc:105: rocksdb::Status rocksdb::CacheReservationManagerImpl<R>::ReleaseCacheReservation(std::size_t) [with rocksdb::CacheEntryRole R = rocksdb::CacheEntryRole::kBlockBasedTableReader; std::size_t = long unsigned int]: Assertion `GetTotalMemoryUsed() >= incremental_memory_used' failed.
Received signal 6 (Aborted)
```

hx235 replied:
Let me look into that and see if we should pursue another way of walking through TableProperties

pdillinger replied:
The alternative it to iterate up to and including the last `std::string*` given by the (current) last std::string field. That should be safe as long as all the fields in between are std::string. The only downside is that adding another string field at the end of the list will require updating the code.

hx235 replied:
Update: for the purpose of approximating memory, I decided to explicitly enumerating std::string fields of TableProperties for readability as both iteration hacks and explicit enumeration suffer from inflexibility in adding new string prop.

For the purpose of existing TEST_SetRandomTableProperties, since it already exists and requires mutation of field, I opened another PR to fix the issue caused by potential non-zero padding https://github.com/facebook/rocksdb/pull/9812

Will merge this PR after CI is clear. Thanks!

ajkr reviewed (commented):
LGTM except I need to figure out what happened in table_properties.cc

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -13,54 +13,96 @@
-template <CacheEntryRole R>
-class CacheReservationHandle;
+ protected:
+  virtual Status ReleaseCacheReservation(
+      std::size_t /* incremental_memory_used */) {
+    return Status::NotSupported();
+  }
```
Nothing bad happens - I just forgot to remove it.

hx235 replied:
Done

hx235 commented on cache/cache_reservation_manager.h:
```diff
@@ -172,20 +214,81 @@ class CacheReservationManager
+    }
+    (*handle).reset(
+        new ConcurrentCacheReservationManager::CacheReservationHandle(
+            std::enable_shared_from_this<
```
It seems that ConcurrentCacheReservationManager is a subclass of public std::enable_shared_from_this<ConcurrentCacheReservationManager> https://github.com/facebook/rocksdb/pull/9748/files#diff-b95042a5d09f06814f920b8c334a4f0aff0574c41981d1fc82a93d1c87e774b5R219 ? Did I misunderstand your question?

ajkr replied:
You're right, I was confused.

ajkr reviewed (approved):
LGTM!

ajkr commented on cache/cache_reservation_manager.cc:
```diff
@@ -66,41 +85,32 @@ Status CacheReservationManager::UpdateCacheReservation(
+      UpdateCacheReservation(GetTotalMemoryUsed() + incremental_memory_used);
+  (*handle).reset(new CacheReservationManagerImpl::CacheReservationHandle(
+      incremental_memory_used,
+      std::enable_shared_from_this<
```
Disregard. I kept getting confused by enable_shared_from_this because it's named like a function...

hx235 replied:
In the latest impl, enable_shared_from_this is moved down subclasses level.

If we keep it at CacheReservationManager level, we will have 
```
std::static_pointer_cast<ConcurrentCacheReservationManager>(std::enable_shared_from_this<
                CacheReservationManager>::shared_from_this())
``` 

One downside is that CacheReservationManager means to be an interface (somewhat abstract away from impl) while `std::enable_shared_from_this` is motivated by the specific implementation of the handle (i.e, pointing back to the CRM), which is what I realized from https://github.com/facebook/rocksdb/pull/9748#discussion_r841248234.

One upside is that compiler can easily catch `std::unique_ptr<CacheReservationManager>` being wrong.

I am good with either ways!
